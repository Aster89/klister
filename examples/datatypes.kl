#lang kernel

(datatype (Nat)
  (zero)
  (add1 (Nat)))

(datatype (List A)
  (nil)
  (:: A (List A)))

(datatype (Tree A)
  (leaf)
  (branch A (Tree A) A))

(datatype (Either A B)
  (left A)
  (right B))

(define nothing (zero))

(define something (add1 (add1 (zero))))

(example (left (branch (zero) (leaf) (add1 (zero)))))
(example (right (right (right (nil)))))


(define double
  (flet (f (n)
          (case n
            [(zero) (zero)]
            [(add1 k) (add1 (add1 (f k)))]))
    f))

(example (double (add1 (add1 (zero)))))

(define flop
  (lambda (either)
    (case either
      [(left a) (right a)]
      [(right b) (left b)])))

(example (flop (left #true)))
(example (flop (right (:: (add1 (zero)) (nil)))))

(define plus
  (flet (f (n)
          (case n
            [(zero) (lambda (k) k)]
            [(add1 n*) (lambda (k) ((f n*) (add1 k)))]))
    f))

(example ((plus (add1 (add1 (zero)))) (add1 (add1 (add1 (zero))))))

(export Nat zero add1)
