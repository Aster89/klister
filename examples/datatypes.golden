(left (branch (branch (leaf) (zero) (leaf)) (add1 (zero)) (leaf))) : ∀α0. (Either (Tree Nat) α0)
(right (right (right (nil)))) : ∀α0 β0 γ0 δ0. (Either α0 (Either β0 (Either γ0 (List δ0))))
(add1 (add1 (add1 (add1 (zero))))) : Nat
(right #true) : ∀α0. (Either α0 Bool)
(left (:: (add1 (zero)) (nil))) : ∀α0. (Either (List Nat) α0)
(add1 (add1 (add1 (add1 (add1 (zero)))))) : Nat
